Added test scripts for TC_SEPR_2.1/2/3.py

jamesharrow opened the pull request:
Splitting the Python Test scripts into separate PR to allow Python experts to review without the SDK changes.

This PR assumes that the recent TP PR (https://github.com/CHIP-Specifications/chip-test-plans/pull/5066) is merged.
That in turn assumes that the spec PR (https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/11569) is merged.

#### Testing

Testing with CI running tests.
[Previous issues were observed #38580  and #38575 have now been fixed. This PR has merged in #38622 - but depending on merge order will disappear].

jamesharrow committed:
Added basic test steps for TC_SEPR_2.1/2/3.py

jamesharrow committed:
Marked TC_SEPRTestBase as shared code (not a script)

jamesharrow committed:
Merge remote-tracking branch 'upstream/master' into TC_SEPR_test_scripts

jamesharrow committed:
Cherry picked test scripts from energy-gateyway-app-2 to split into smaller review

restyled-commits committed:
Restyled by isort

jamesharrow committed:
Fix Lint error

jamesharrow committed:
Updated TC_SEPR_2_3 to remove ForecastChange event checking

jamesharrow committed:
Ruff fixes

jamesharrow committed:
Fixed test step 6 missing.

andy31415 committed:
Merge branch 'master' into TC_SEPR_test_scripts

andy31415 commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,195 @@
+                                             cluster: Clusters.CommodityPrice = None,
+                                             struct: Clusters.CommodityPrice.Structs.CommodityPriceStruct = None,
+                                             details: Clusters.CommodityPrice.Bitmaps.CommodityPriceDetailBitmap = 0,
+                                             check_time: bool = True):
```
check_time feels vague. It seems that this is a `now_time_must_be_within_period` ... although that feels a bit verbose.

Maybe a doccomment explaining things would help.

jamesharrow replied:
Added doc string and changed to suggested name

andy31415 commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,195 @@
+                                             endpoint: int = None,
+                                             cluster: Clusters.CommodityPrice = None,
+                                             struct: Clusters.CommodityPrice.Structs.CommodityPriceStruct = None,
+                                             details: Clusters.CommodityPrice.Bitmaps.CommodityPriceDetailBitmap = 0,
```
is 0 a good initializes for the bitmap? There is no type casting in python so I wonder if trying to get `0.kDescription` would fail. I think this should be either defaulting to an empty bitmap or be optional and initialized with None (and code should check for it)

andy31415 replied:
From the usage we do `details`  as a check, so I guess we use 0 as a falsy value...that does not seem ideal. Lets use none or an empty bitmap as default (empty bitmap seems to lead to more readable code)

jamesharrow replied:
Used Clusters.CommodityPrice.Bitmaps.CommodityPriceDetailBitmap(0) - thanks

andy31415 commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,195 @@
+            return result
+
+        except InteractionModelError as e:
+            asserts.assert_equal(e.status, expected_status,
```
Should we assert that status is not success? It seems odd if we execute things, we get a IM error exception and accept it because status is success...

jamesharrow replied:
Added

jamesharrow replied:
Now my head hurts :) we don't want an exception, but this function could expect one (it's a clone from other DEM tests where we expected some negative tests and wanted to check we got the right error code). Here though we mostly want success. So the try could asset it IS a Success (not ! Success).

I think it is ok?

andy31415 replied:
Ok for the try/catch I believe, but see my comment above: on the success path, we have to assert that we are actually expecting success, because success does not go through the exception path.

cecille replied:
OK, so if you're expecting success all the time, you can just remove the try/catch, right? The send command throws on error, which causes a test failure.

I think the addition of the expected_status was what was throwing us.

jamesharrow replied:
Yes - removed all of that and the expect_status nonsense.

andy31415 commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,195 @@
+        logger.info(
+            f"  Component: price: {struct.price} source: {struct.source}, desc: {struct.description} tariffComponentID: {struct.tariffComponentID}")
+
+    async def send_get_detailed_price_request(self, endpoint: int = None, details: Clusters.CommodityPrice.Bitmaps = 0,
```
0 is not a `Bitmaps` and `None` is not a `int`. I think these should probably be required and typings fixed.

jamesharrow replied:
Changed

andy31415 commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,195 @@
+                                 "Unexpected error returned")
+
+    async def send_test_event_trigger_price_update(self):
+        await self.send_test_event_triggers(eventTrigger=0x0095000000000000)
```
add named constants for these magic values

andy31415 replied:
also should we not be using `self.matter_test_config.global_test_params['PIXIT.SEPR.TEST_EVENT_TRIGGER_KEY'] ` constants and similar instead of hardcoding values?

I am looking at examples in https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/TC_BOOLCFG_4_2.py#L93

jamesharrow replied:
Not for these. Test Events have a unique value for the cluster to make some test plan spec defined action happen, the 'Key' is a manufacturer defined secret that they can change in their 'test' firmware.

Other clusters (ok that I've written like TC_DEMTestBase.py / TC_EEVSE_Utils.py all do it this way.

I've added constants though for reability.

andy31415 commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,195 @@
+
+            return matter_epoch + delta_from_epoch
+        else:
+            return "None"
```
Return type is odd: string or datetime. Should this return `None` instead (not a string)?

jamesharrow commented on src/python_testing/TC_SEPR_2_1.py:
```diff
@@ -0,0 +1,138 @@
+                cluster.Structs.CommodityPriceStruct, allow_empty=True)
+            for item in val:
+                # In the PriceForecast attribute we must not have Description or Components in this returned list
+                await self.test_checkCommodityPriceStruct(endpoint=endpoint, cluster=cluster, struct=item, details=0)
```
Set to Clusters.CommodityPrice.Bitmaps.CommodityPriceDetailBitmap(0) to ensure it is matching the comment here

jamesharrow committed:
Updates following review - and adding force TCP connection for TC_SEPR_2_3

restyled-commits committed:
Restyled by isort

jamesharrow commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,195 @@
+class CommodityPriceTestBaseHelper:
+
+    async def test_priceForecast(self,
+                                 endpoint: int = None,
```
I inherited this from Alchemy autogen test script (but maybe I mangled it somewhat). Agree it should be tidied. Perhaps as a follow up after TE2 FC? I'm unlikely to be reliably able to track the other 5 PRs and do this without breaking this script.

jamesharrow replied:
Have now tidied this up.

jamesharrow committed:
Fixed ruff issues

restyled-commits committed:
Restyled by isort

cecille commented on src/python_testing/TC_SEPR_2_3.py:
```diff
@@ -0,0 +1,221 @@
+
+        return steps
+
+    @async_test_body
```
You can use run_if_endpoint_matches here too - just use has_feature.

jamesharrow replied:
Good to know - updated

cecille commented on src/python_testing/TC_SEPR_2_3.py:
```diff
@@ -0,0 +1,221 @@
+
+    @async_test_body
+    async def test_TC_SEPR_2_3(self):
+        # pylint: disable=too-many-locals, too-many-statements
```
...why is the linter breaking here? This doesn't seem like too many locals or statements at all.

jamesharrow replied:
Copied from old script - removing.

jamesharrow committed:
Merge branch 'master' into TC_SEPR_test_scripts

jamesharrow commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,225 @@
+        else:
+            asserts.assert_is_none(struct.components)
+
+    async def test_checkCurrencyStruct(self,
```
Ah good to know - again inherited from Alchemy autogen (@hasty FYI)

jamesharrow replied:
Updated

jamesharrow committed:
Apply suggestions from code review

Co-authored-by: C Freeman <cecille@google.com>

jamesharrow committed:
Updates based on review comments; removed async, renamed funcs to avoid test_ and endpoint:int = None

jamesharrow commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,225 @@
+    async def send_test_event_trigger_forecast_update(self):
+        await self.send_test_event_triggers(eventTrigger=self.kEventTriggerForecastUpdate)
+
+    def get_current_time_as_epoch_s(self):
```
Slightly different since that returns *100000, but refactored to use it and then divide by 100000 again!

jamesharrow committed:
removed get_current_time_as_epoch_s

jamesharrow committed:
Changed async_test_body to run_if_endpoint_matches with Forecasting feature.
Removed pylint disable

jamesharrow committed:
Disabled TC_SEPR_2_3 which fails because of #38575

restyled-commits committed:
Restyled by isort

andy31415 commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,209 @@
+                endpoint=endpoint,
+                timedRequestTimeoutMs=timedRequestTimeoutMs)
+
+            return result
```
if expected_status is not success but we succeed anyway this would pass.

Lets assert here that expected_status is success (and for other places as well)

jamesharrow replied:
Done

andy31415 commented on src/python_testing/test_metadata.yaml:
```diff
@@ -52,6 +52,12 @@ not_automated:
+    - name: TC_SEPR_2_3.py
+      reason:
+          Bug in chip-repl for TCP handling make this fail.
+          https://github.com/project-chip/connectedhomeip/issues/38575
```
can we just fix the bug? It seems that changing src/controller/python/chip/clusters/command.cpp is enough according to that issue.

andy31415 replied:
I have concerns merging a test that is known to fail in the current SDK (so it will not be maintained by CI).

I would split this test out and depend the merging of it on fixing #38575

jamesharrow replied:
The python test script is fine, and I'd prefer to keep the python code together in a PR if we can...

I'm not sure that the fix to make the buffer bigger was what @bzbarsky-apple was thinking, instead having proper handling based on the type of connection (TCP vs UDP).
I 'can' apply that larger buffer change if you think it's the best short-term solution, but don't want to break something else.

And there is a potential other side (server) issue I have also raised which I don't know how to fix - but will keep looking at - https://github.com/project-chip/connectedhomeip/issues/38580). I was proposing we re-enable this in the CI once both bugs are fixed (since in theory the python code doesn't need to change and getting python code reviews seems to be hard!). 

Changing a yaml file to delete 2 lines is easy for people to approve.

cecille replied:
The TH is built off the SDK though, so won't this mean the test will fail at the test event if you don't get the change in?

jamesharrow replied:
I'd rather it fails at test event and it encourages us to fix it, than not ship the 2.3 test case... then it raises its awareness that we need to make the Forecast feature provisional in 1.5 if we don't get that working... but its sort of the whole point of the cluster is to get a forecast! The current price is interesting, the future prices tells a home how to best use energy in the near future.

i.e. I'm really keen to get this fix in in the next few days. The challenge is that I'm not sure I'm the right person to fix the TCP stuff elegantly. I can put the hack in, but would rather do that in a separate PR. but the TE2 cutoff is tonight. Better to have 2 out of 3 test cases passing and one known issue which I'll be chased on.

jamesharrow replied:
There is a PR for the larger buffer fix for chip-repl (for test harness), https://github.com/project-chip/connectedhomeip/pull/38607

but unfortunately there is also the server side crash which happens when handling the command over TCP. So this will still block TC_SEPR_2_3 from being runnable -

jamesharrow replied:
Good news - Boris has fixed #38580.
I've pulled that fix into this branch, and enabled TC_SEPR_2_3 again - to prove both work in CI.

jamesharrow committed:
Removed expected_status which isn't required

jamesharrow committed:
Merge branch 'master' into TC_SEPR_test_scripts

jamesharrow committed:
Added missing build command to build the app!

jamesharrow committed:
Merge branch 'upstream-master' into TC_SEPR_test_scripts

jamesharrow committed:
Commented out unused variables which causes compilation failure on clang-asan builds.

jamesharrow committed:
Applied changes suggested from review - removed args which aren't really needed

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by isort

bzbarsky-apple committed:
Fix crash when a timed invoke happens for a large payload command.

Because we don't pass the exchange through command handling, treating it as
out-of-band state instead, we have to maintain that state on all codepaths that
can lead to command handling.

Also fixes timed writes, but those are not exercised right now.

Fixes https://github.com/project-chip/connectedhomeip/issues/38580

jamesharrow committed:
Merged in TC_EGCTestBase.p (to avoid merge conflicts), Removed TC_SEPR_2_3.py being disabled now we have a fix for Large TCP commands.

jamesharrow committed:
Merge branch 'master' into TC_SEPR_test_scripts

andy31415 commented on src/python_testing/TC_SEPRTestBase.py:
```diff
@@ -0,0 +1,190 @@
+
+        if now_time_must_be_within_period:  # Only check time limits when dealing with current price (not list of Forecast)
+            # - verify that the PeriodStart is in the past.
+            now_time_epoch_s = utc_time_in_matter_epoch() // 1000000
```
```suggestion
            now_time_epoch_s = utc_time_in_matter_epoch() // 1_000_000
```

makes it easier to read

andy31415 commented on src/python_testing/test_metadata.yaml:
```diff
@@ -24,6 +24,8 @@ not_automated:
       reason: Shared code for TC_EWATERHTR, not a standalone test
     - name: TC_EnergyReporting_Utils.py
       reason: Shared code for TC_EEM and TC_EPM, not a standalone test
+    - name: TC_EGCTestBase.py
```
Technically not part of this PR

jamesharrow replied:
I did this to avoid another 3hour CI run in case the other PR merged first.

jamesharrow committed:
Update src/python_testing/TC_SEPRTestBase.py

Co-authored-by: Andrei Litvin <andy314@gmail.com>

jamesharrow committed:
Merge branch 'master' into TC_SEPR_test_scripts
